DNA project

muffinmad opened the pull request.

muffinmad committed:
DNA project

Svyatoslavik commented on madDNA/madDNA/Cell.m:
```diff
@@ -0,0 +1,46 @@
+}
+
+- (void)mutateAtPosition:(int)index {
+	[_DNA replaceObjectAtIndex:index withObject:[self getRandomChar]];
```
Данный код не гарантирует того что символ будет замещен на другой - в результате чего не произойдет мутации
например у вас на даной позиции находиться символ "A" и метод [self getRandomChar] так же возвращает "A".

Svyatoslavik commented on madDNA/madDNA/main.m:
```diff
@@ -0,0 +1,47 @@
+@implementation Cell (mutator)
+
+- (void)mutate:(int)percent {
+	NSMutableSet *mutated = [NSMutableSet set];
```
Было бы хорошо вынести категорию в отдельный файл

muffinmad committed:
изменения в замене символа и в выборе символов для замены

- Теперь новый символ при замене должен отличаться от старого
- Механизм выбора индекса для замены больше не ждет пока случай
предоставит ему индекс символа, ранее не заменявшегося

muffinmad committed:
категория теперь в отдельных файлах

muffinmad committed:
readonly свойство заменил все же переменной, как указано в задании

Svyatoslavik commented on madDNA/madDNA/Cell.m:
```diff
@@ -0,0 +1,56 @@
+	NSString *current_char = [self charAtPosition:index];
+	NSString *new_char = [self getRandomChar];
+	while (current_char == new_char) {
+		new_char = [self getRandomChar];
```
попахивает индуским кодом
плохой метод замены - получаеться что вы очень сильно привязаны к генератору случайных чисел - допустим он заглючит (маловероятно конечно но все же) тогда и ваша програма зациклиться
суть такова что нужно текущий символ вообще исключить из возможных случайных вариантов
ведь кто то может сделать и так
// file main.m
u_int32_t arc4random(void){
    return 0;
}
ваша праграма не должна быть зависима от реализации генератора случайных чисел

muffinmad committed:
изменения в выборе нового символа

Теперь новый символ выбирается из доступных
+ немного коментариев
